### PR TITLE
ci: Upload coverage from each job

### PR DESCRIPTION
**Pull request by dnephin**
Allows us to remove the merge coverage report step.

### CODE REVIEW

**Comment by dnephin** [comment]
This seems to work: https://codecov.io/gh/hashicorp/consul/commit/44525041993a2136fb9d82d332bd6a754bf47150/build

File: .circleci/config.yml [hunk 1]
```diff
@@ -174,19 +168,14 @@ jobs:
...
-            - cov_*.part
-
       - store_test_results:
           path: *TEST_RESULTS_DIR
       - store_artifacts:
           path: *TEST_RESULTS_DIR
+      - run: *codecov_upload
 
   # combine code coverage results from the parallel circleci executors
   coverage-merge:
```
Comment (alvin-huang) [inline .circleci/config.yml]: Can you remove this whole job now that we don't need it?

Comment (alvin-huang) [reply .circleci/config.yml]: Also can you add back  `go tool cover -html=coverage.out -o $TEST_RESULTS_DIR/coverage.html`? I think this is helpful if you don't want to use codecov but just click on the HTML in the CircleCI artifacts to get the normal Go built in HTML explorer of coverage

Comment (dnephin) [reply .circleci/config.yml]: Good catch, I forgot to remove the `coverage-merge` job. Did that with an amend.

Now that we upload from 6 different tasks the html report probably isn't that useful. I suspect the html report is not used all that often, and when it is , we could probably generate it just as easily by running the tests locally.

File: codecov.yml [hunk 2]
```diff
@@ -30,3 +30,8 @@ comment: false
 # https://docs.codecov.io/docs/flags
 # TODO: split out test coverage for API, SDK, UI, website?
 # flags:
+
+ignore:
+  - "agent/bindata_assetfs.go"
+  - "vendor/**/*"
```
Comment (alvin-huang) [inline codecov.yml]: This is fine to have but isn't needed since we don't push coverage reports from this folder up so it wouldn't show up anyway.

**Comment by dnephin** [comment]
Pushed one addition, I added `when: always` to the upload step, so that we get coverage reports even when tests flake.

**Comment by alvin-huang** [comment]
> Pushed one addition, I added `when: always` to the upload step, so that we get coverage reports even when tests flake.

Hmm, what actually happens to the go test output if it errors/OOMs/doesn't finish? Does the coverage report still get generated?

**Comment by dnephin** [comment]
> if it errors/OOMs/doesn't finish, Does the coverage report still get generated?

I'm not sure what happens in the OOM or SIGKILL case. I would guess that it writes the output as a stream so the file is probably partially populated. If a file does not exist then the upload will fail. In those cases I suspect it's fine that our coverage report is under reporting, because the test will have failed and will need to be re-run anyway.

If the tests fail with an error it should produce a normal coverage report, but I haven't tested this specifically.

**Pull request merged**